### ISSUE TITLE
New `max-lines-per-function` rule

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#9842]
**Please describe what the rule should do:** It should allow a configurable limit of the max number of lines per function, much like how `max-statements` limits the number of statements. Also, there should be an `"ignoreComments"` boolean option, on by default, that skips comments when counting method length.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// eslint max-lines-per-function: ["error", 1]
// Good
function foo() { return 1 }
const bar = () => 2

// Bad
function foo() {
    return 1
}
const bar = () =>
    2

// eslint max-lines-per-function: ["error", 5, {ignoreComments: false}]
// Good
function foo() {
    // This is a multi-line
    // comment.
    doSomething();
}

// Bad
function foo() {
    // This is a multi-line
    // comment.
    doSomething();
    doSomethingElse();
}

function foo() {
    // This is a multi-line
    // comment that takes
    // three lines.
    doSomething();
}

// eslint max-lines-per-function: ["error", 3, {ignoreComments: true}]
// Good
function foo() {
    // This is a multi-line
    // comment.
    doSomething();
}

// Bad
function foo() {
    // This is a multi-line
    // comment.
    doSomething();
    doSomethingElse();
}

function foo() {
    doSomething();
    doSomethingElse();
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It is a simple, basic rule similar to what already exists, and is a natural extension of a few of them. It's also very widely generic, and although it could replace `max-statements` in some uses, it checks a different metric. It also does *not* conflict with any other rules, and it's pretty much independent from everything short of the language itself.

Other linters and style guides in various languages already do similar:

- Rubocop has [`Metrics/MethodLength`](http://rubocop.readthedocs.io/en/latest/cops_metrics/#metricsmethodlength) which does exactly what I propose here, including the proposed option.
- Google's JS style guide ([by proxy](https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Code_formatting) of their [C++ style guide](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions)), recommends 40 lines or fewer per function.
- [The Linux Kernel's style guide](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#functions) recommends 48 lines or fewer.

<details>
<summary>This is the original bug I filed</summary>

It should basically allow a configurable limit of the max number of lines per function, much like how `max-statements` limits the number of statements. This has precedent: Google's JS style guide ([by proxy](https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Code_formatting) of their [C++ style guide](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions)), recommends 40 lines or fewer per function. Also, to draw another language's style guide, [the Linux Kernel](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#functions) recommends 48 lines or fewer.
</details>

**Comment by not-an-aardvark** [comment]
Closing because not enough information was provided to triage this issue.

**Comment by isiahmeadows** [comment]
@not-an-aardvark What other information is needed?

**Comment by platinumazure** [comment]
Hi @isiahmeadows, it would really help us to have the information requested in this set of guidelines: [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules). Thanks!

**Comment by isiahmeadows** [comment]
@platinumazure I updated the original bug to include all that info.

**Comment by platinumazure** [comment]
@isiahmeadows Thank you! I'll go ahead and reopen this issue.

**Comment by platinumazure** [comment]
@isiahmeadows Very sorry for the long delay on this. Just wanted to ask, why is max-statements not sufficient for this sort of thing? (I realize that the counting is different-- statements vs lines-- but I'm asking why it's insufficient, not why it's different.)

I think I can probably get behind this, but I want to make sure there's a compelling case for implementing a line-based rule here when we should already have one for statements.

**Comment by isiahmeadows** [comment]
@platinumazure Some expressions (think: long method chains, JSX/hyperscript trees, functional pipelines) could take several lines, and `max-statements` would see only a few statements.

For a concrete example, in [this method](https://github.com/MithrilJS/mithril.js/blob/next/examples/dbmonster/mithril/app.js#L10-L34):

- `max-statements` would see exactly one statement, even though it's 25 lines long.
- `max-depth` would see no nested blocks, so it wouldn't be a limiting factor here, either.
- `max-nested-callbacks` would see a depth of at most 3, which is much lower than what most care about (default is 10, but I keep it at 4 for my personal projects).
- `complexity` is 1, so it has little-to-no use here.

Another concrete example, [this method](https://github.com/MithrilJS/mithril.js/blob/next/examples/todomvc/todomvc.js#L89-L125):

- `max-statements` would only see two statements, even though it's 36 lines long (a large method by any standard).
- `max-depth` would see a depth of at most 1 - there's no blocks, just calls and function expressions.
- `max-nested-callbacks` would see a depth of at most 3.
- The only rule that would have much of an effect limiting it would be `complexity`, which would rate the outermost at 7, the `.map` callback at 3, and the rest at 1 (default is 10).

So there's uses for it, just in the more pathological cases. (Of course, most code doesn't have this issue.)

**Comment by platinumazure** [comment]
Alright, thanks, I'm sold. TBH I'm not going to push hard for this with the team right now since we're doing major release candidates, but I'll champion this after we finish that off. (That's a promise but real life does get in the way- feel free to ping me if you think I might need a reminder. Thanks!)

**Comment by platinumazure** [comment]
As promised, I'll champion this.

**Comment by platinumazure** [comment]
@eslint/eslint-team Could we get two more :+1:s on this? This seems like a worthwhile addition since it complements other "max" rules which enforce readability.

**Comment by mysticatea** [comment]
How does it behave in the following case?

```js
function foo(
    aaa = 1,
    bbb = 2,
    ccc = 3,
) {
    return aaa + bbb + ccc
}
```

I guess that the rule handles function head and body separately.

**Comment by peteward44** [comment]
In your example it will report 7 lines (it includes the head). The rule listens to the ```FunctionDeclaration``` node in the AS tree.

**Comment by platinumazure** [comment]
Seems to me like we could start with one approach or the other, and then enhance the rule to support the other use case later if needed. :smile:

**Comment by mysticatea** [comment]
Though I felt natural if it counts only function bodies. But I'm OK if the lines of whole functions is more popular.

**Comment by alberto** [comment]
How would it treat IIFEs and nested functions?

**Comment by peteward44** [comment]
As it works at the moment, nested functions are included in their parent's count. Nested functions are also counted individually (so if a nested function exceeds the limit, both it and it's parent will fail).

For IIFEs the rule will just ignore the surrounding parentheses; so the following will report 4 lines:
```
(
function
()
{
}
)
()
```

Literally all the rule does is look for "function" (or starting parentheses for fat arrow) and the last curly bracket then count the line endings between them.

I will extend the unit tests to include the examples you guys have raised though.

**Comment by mysticatea** [comment]
Ahh, and how about class methods? Actually, a part of class methods is `FunctionExpression`, but some modifiers and method names are not a part of `FunctionExpression`.

I guess currently...:

```js
class A {
    static
    foo
    // This FunctionExpression starts from below `(`, so this is 3.
    (a) {
        return a
    }
}

// Getters/setters are similar to it.
var obj = {
    get
    foo
    // This FunctionExpression starts from below `(`, so this is 3.
    () {
        return 1
    }
}

// The computed property cases can be longer.
class A {
    static
    [
        foo +
            bar
    ]
    // This FunctionExpression starts from below `(`, so this is 3.
    (a) {
        return a
    }
}
```

I think more consistent if this rule behaves either:

1. ignores methods or 
2. counts lines from the top of `Property`/`MethodDefinition` nodes.

**Comment by peteward44** [comment]
Yep, i've just tested those examples and you are right. Ignoring methods seems to defeat the point of the rule so i'm going to try your other suggestion.

**Comment by not-an-aardvark** [comment]
To me, it seems like it could be a usability issue if the rule always reports IIFEs. Wrapping a bunch of function definitions in an IIFE is fairly common to avoid polluting the global scope, and it doesn't make the code significantly more complex, so it seems odd for the rule to report an error for that.

Would it make sense to count the number of lines for which a function is the "most nested" function? In other words, if a statement appears inside a function which is itself inside another function, the line would only count towards the total of the inner function.

**Comment by ljharb** [comment]
In a world with modules, IIFEs are obsolete, so i doubt its actually common anymore.

**Comment by platinumazure** [comment]
@ljharb Never underestimate the fraction of people stuck on ES5 (or ES3 for that matter). :cry:

**Comment by ljharb** [comment]
Sure, so maybe we put IIFEs behind a separate option on the rule.

**Comment by mysticatea** [comment]
@not-an-aardvark I agree about IIFEs. I think difficult that the rule considers callback functions. For example, I felt that the callbacks of `Array#forEach` should be a part of the parent function, but I felt that the callback of event listeners should not increase the lines of the parent function. It's difficult.

**Comment by ljharb** [comment]
The rule is “per function”, and i suspect the intent - and only thing that makes sense to me - is that every function is considered entirely on its own. How nested a function is doesn’t make it more or less OK to have a certain number of lines, and shouldn’t change the allowed amount.

**Comment by peteward44** [comment]
I've added an ```ignoreIIFEs``` option. It defaults to true; as I figure anyone who doesn't use them won't care; and anyone who does use them will almost certainly not want max-lines-per-function to count the lines in them.

**Comment by ljharb** [comment]
Having boolean options that don't default to `false` is super weird; although I'd prefer `ignoreIIFEs` to default to `false`, if you want the default to be "ignore IIFEs", then can we pick an option name that makes sense when defaulting to `false`?

**Comment by platinumazure** [comment]
I think @ljharb is suggesting that it might be better to have an "opt in" option name such as `enforceForIIFEs`. I kind of agree.

**Comment by ljharb** [comment]
Yes - basically that for any boolean option, absence and false should be the same, for maximum clarity.

**Comment by peteward44** [comment]
OK i've updated the PR with all the feedback so far. Thanks all. As far as whether IIFEs should be included by default or not i'll go with whatever you guys want; but i agree with @not-an-aardvark that it would cause usability issues, and i'm of the opinion that default values should generally be the most user-friendly possible 🤷‍♂️

**Comment by ljharb** [comment]
My opinion is that defaults should be as restrictive as possible as well, so users are loosening them with intention.

**Comment by platinumazure** [comment]
Now that we have a champion and three :+1: from the team, this is accepted.

**Comment by Mouvedia** [comment]
#6146
